Make the code more sync

warpech opened the pull request:
Alternative solution to PR #9

warpech committed:
Problem: there was a problem explained in https://github.com/Juicy/juicy-select/pull/9, in which `juicy-select` performs an async operation even after it was detached from DOM

Solution: make the code more sync

miyconst reviewed (changes requested):
@warpech unfortunately I can't accept this pull request.

The `<juicy-select>` was explicitly developed to keep the `<select>` element in the light DOM. 

After your rework the following construction will no longer work:

```html
<juicy-select>
    <select class="custom-class" on-change="onSelectChangeCustom"></select>
</juicy-select>
```

warpech commented:
Thanks, now I understand the motivation.

This is quite against the common practice in custom elements. Normally, we would do it like this then:

```html
<juicy-select-bind-options options="{{options}}">
    <select slot="select" class="custom-class" on-change="onSelectChangeCustom"></select>
</juicy-select-bind--ptions>
```

I will make a change.

miyconst commented:
>This is quite against the common practice in custom elements. Normally, we would do it like this then:

Could be. This is how the `pikaday` decorator works, and it's very nice. The element may be renamed, but renaming itself won't solve the issue.

Note, that `<juicy-select>` allows a `<select>` in it's light DOM, but does not require it. If no `<select>` element provided, one will be created.

warpech committed:
Problem: .gitignore file is specific for .Net, not for web components. It does not include bower_components, node_modules

Solution: replace .gitignore with one that includes bower_components, node_modules

warpech committed:
Problem: the demo page does not display the selection in multiselect

Solution: dom-if and dom-repeat were used incorrectly. They should be used as <template> tag, similarly to Polymer 1

warpech committed:
Problem: 22d4db4 introduced a breaking change - moved <select> from light DOM to Shadow DOM

Solution: revert to light DOM, protect methods from executing until options are assigned

warpech commented:
Could you please re-review?

---

> Could be. This is how the `pikaday` decorator works, and it's very nice. The element may be renamed, but renaming itself won't solve the issue.

Yes, I mean this as a potential new custom element. But we don't have resources to work on it now.

miyconst commented on juicy-select.html:
```diff
@@ -27,45 +27,41 @@
                 }
             },
             generateOptions: function () {
+                if (!this.options || !this.options.length) {
```
In this case the empty option or caption text will not be shown when there are no options available, but it should be.

warpech replied:
Fixed. I thought this is not intended use case, but I agree I should not make such breaking changes.

warpech committed:
Problem: in ceb585d, checking for options array was moved to the top of the method, causing a breaking change.(the empty option or caption text will not be shown when there are no options available, but it should be)

Solution: revert this check to the previous position

eriksunsol commented:
As an alternative solution to PR #9, this does not work. See https://github.com/Juicy/juicy-select/pull/9#issuecomment-341411932.

warpech committed:
Problem: the demo page does not have an example of using a provided light DOM select

Solution: add a demo for that

alshakero committed:
Enhancements

1. Add a few tests
2. Enable Travis
3. Fix #9

warpech commented on juicy-select.html:
```diff
@@ -78,8 +83,12 @@
 
                     this.$select.appendChild(opt);
                 }
+                console.log(this.$select.childNodes.length)
```
Please remove `console.log`

warpech commented on juicy-select.html:
```diff
@@ -38,19 +38,24 @@
+                if (this.options.length && this.$select) {
+                    this.generateOptions();
+                }
+                this.addEventListener("change", this.selectChange);
```
Do you need both event listeners, one on line 41 and one on line 49?

alshakero replied:
Good catch.

warpech commented on test/helpers.js:
```diff
@@ -0,0 +1,7 @@
+function simulateChangeEvent(element) {
+  if ('createEvent' in document) {
+    var evt = document.createEvent('HTMLEvents');
+    evt.initEvent('change', false, true);
```
`initEvent` is [deprecated](https://developer.mozilla.org/en-US/docs/Web/API/Event/initEvent).

Pls use something like:

```js
var evt = new Event('change', {bubbles: true, composed: true});
element.dispatchEvent(evt);
```

warpech commented on test/simple.html:
```diff
@@ -0,0 +1,132 @@
+                const childNodes = Object.entries(juicySelect.$select.childNodes);
+
+                for (let [index, option] of childNodes) {
+                    index = ~~index;
```
Please don't use bitwise sorcery. 🧙 Do you [mean Math.floor](https://stackoverflow.com/questions/5971645/what-is-the-double-tilde-operator-in-javascript)?

alshakero replied:
It's a fast alternative to parseInt. But you're right.

alshakero committed:
Modernize event simulation and drop for...of

alshakero commented:
I'll merge and release a minor.

alshakero commented:
Released here and on NPM.

warpech commented:
Setting as “Finished” because it is:
- released (with release notes) https://github.com/Juicy/juicy-select/releases/tag/1.1.0
